mitiq/pec/representations/optimal.py

```diff
@@ -0,0 +1,130 @@
+# Copyright (C) 2020 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Functions for finding optimal representations given a noisy basis."""
+
+from typing import List
+from mitiq import QPROGRAM
+from mitiq.pec.types import NoisyBasis, OperationRepresentation
+import numpy as np
+from scipy.optimize import minimize, LinearConstraint
+from mitiq.pec.channels import matrix_to_vector, kraus_to_super
+from cirq import channel
+from mitiq.conversions import convert_to_mitiq
+
+
+def minimize_one_norm(
+    ideal_matrix: np.ndarray,
+    basis_matrices: List[np.ndarray],
+    tol: float = 1.0e-8,
+) -> List[float]:
+    r"""
+    Returns the list of real coefficients :math:`[x_0, x_1, \dots]`,
+    which minimizes :math:`\sum_j |x_j|` with the contraint that
+    the following representation of the input ``ideal_matrix`` holds:
+
+    .. math::
+        \text{ideal_matrix} = x_0 A_0 + x_1 A_1 + ...,
+
+    where :math:`\{A_j\}` are the basis matrices, i.e., the elements of
+    the input ``basis_matrices``.
+
+    This function can be used to compute the optimal representation
+    of an ideal superoperator (or Choi state) as a linear
+    combination of real noisy superoperators (or Choi states).
+
+    Args:
+        ideal_matrix: The ideal matrix to represent.
+        basis_matrices: The list of basis matrices.
+        tol: The error tolerance for each matrix element
+            of the represented matrix.
+
+    Returns:
+        The list of optimal coefficients :math:`[x_0, x_1, \dots]`.
+    """
+
+    # Map complex matrices to extended real matrices
+    ideal_matrix_real = np.hstack(
+        (np.real(ideal_matrix), np.imag(ideal_matrix))
+    )
+    basis_matrices_real = [
+        np.hstack((np.real(mat), np.imag(mat))) for mat in basis_matrices
+    ]
+
+    # Express the representation constraint written in the docstring in the
+    # form of a matrix multiplication applied to the x vector: A @ x == b.
+    matrix_a = np.array(
+        [matrix_to_vector(mat) for mat in basis_matrices_real]
+    ).T
+    array_b = matrix_to_vector(ideal_matrix_real)
+
+    constraint = LinearConstraint(matrix_a, lb=array_b - tol, ub=array_b + tol)
+
+    def one_norm(x):
+        return np.linalg.norm(x, 1)
+
+    initial_guess = np.zeros(len(basis_matrices))
```

Comment:
Allow for optional initial guess as input argument?

Reply:
Added here: 5a649f6